src/algebra/continued_fractions/convergents_equiv.lean

```diff
@@ -0,0 +1,409 @@
+/-
+Copyright (c) 2020 Kevin Kappelmann. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Kappelmann
+-/
+import algebra.continued_fractions.continuants_recurrence
+import algebra.continued_fractions.terminated_stable
+import tactic.linarith
+/-!
+# Equivalence of Convergent Computations for gcfs
+
+## Summary
+
+We show the equivalence of the convergents computations (recurrence relation (`convergents`) vs.
+direct evaluation (`convergents'`)) for `gcf`s on linear ordered fields. We follow the proof from
+[hardy1979introduction], Chapter 10.
+
+## Main Theorems
+
+- `generalized_continued_fraction.convergents_eq_convergents'` shows the equivalence under a strict
+positivity restriction on the sequence.
+- `continued_fractions.convergents_eq_convergents'` shows the equivalence for (regular) continued
+fractions.
+
+## References
+
+- https://en.wikipedia.org/wiki/Generalized_continued_fraction
+- [Hardy, Godfrey Harold and Wright, Edward Maitland, and others,
+*An introduction to the theory of numbers*][hardy1979introduction]
+
+## Tags
+
+fractions, recurrence, equivalence
+-/
+
+variables {α : Type*} {n : ℕ}
+namespace generalized_continued_fraction
+open generalized_continued_fraction as gcf
+variables {g : gcf α} {s : seq $ gcf.pair α}
+
+/-!
+We will show the equivalence of the computations by induction. To make the induction work, we need
+to be able to *squash* the nth and (n + 1)th value of a sequence. This squashing itself and the
+lemmas about it are not very interesting. We hence keep them `private`.
+-/
+
+section squash
+section with_division_ring
+variable [division_ring α]
+
+/--
+Given a sequence of gcf.pairs `s = [(a₀, bₒ), (a₁, b₁), ...]`, `squash_seq s n`
+combines `⟨aₙ, bₙ⟩` and `⟨aₙ₊₁, bₙ₊₁⟩` at position `n` to `⟨aₙ, bₙ + aₙ₊₁ / bₙ₊₁⟩`. For example,
+`squash_seq s 0 = [(a₀, bₒ + a₁ / b₁), (a₁, b₁),...]`.
+If `s.terminated_at (n + 1)`, then `squash_seq s n = s`.
+-/
+private def squash_seq (s : seq $ gcf.pair α) (n : ℕ) : seq (gcf.pair α) :=
```

Comment:
Usually we don't make definitions/lemmas `private` unless one can be completely sure that nobody will ever want to use them (e.g., it's OK to have `private lemma add_comm` that is later used to prove `add_comm_monoid`).